compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/RegExpLexer.java

```diff
@@ -0,0 +1,716 @@
+                token = readTokenInInterpolation();
+                break;
+            default:
+                token = null;
+        }
+
+        return cloneWithDiagnostics(token);
+    }
+
+    /**
+     * Read the next token in the {@link ParserMode#RE_DISJUNCTION} mode.
+     * <p>
+     * <code>
+     * ReDisjunction := ReSequence ("|" ReSequence)*
+     * </code>
+     *
+     * @return Next token
+     */
+    private STToken readTokenInReDisjunction() {
+        reader.mark();
+        if (this.reader.isEOF()) {
+            return getRegExpSyntaxToken(SyntaxKind.EOF_TOKEN);
+        }
+
+        int nextChar = peek();
+        switch (nextChar) {
+            case LexerTerminals.BACKTICK:
+                // End of regular expression.
+                endMode();
+                return nextToken();
+            case LexerTerminals.DOLLAR:
+                if (this.reader.peek(1) == LexerTerminals.OPEN_BRACE) {
+                    // Start interpolation mode.
+                    startMode(ParserMode.INTERPOLATION);
+                    this.reader.advance(2);
+                    return getRegExpSyntaxToken(SyntaxKind.INTERPOLATION_START_TOKEN);
+                }
+                break;
+            case LexerTerminals.PIPE:
+                reader.advance();
+                // Pipe cannot be the end of the regular expression.
+                if (this.reader.isEOF()) {
+                    reportLexerError(DiagnosticErrorCode.ERROR_INVALID_TOKEN_IN_REG_EXP);
+                }
+                return getRegExpText();
+            default:
+                break;
+        }
+
+        // ReSequence has zero or more ReTerm.
+        return readTokenInReTerm();
+    }
+
+    /**
+     * Read token from ReTerm.
+     * <p>
+     * <code>
+     * ReTerm :=
+     *    ReAtom [ReQuantifier]
+     *    | ReAssertion
+     * </code>
+     *
+     * @return Next token
+     */
+    private STToken readTokenInReTerm() {
+        reader.mark();
+        if (reader.isEOF()) {
+            return getRegExpSyntaxToken(SyntaxKind.EOF_TOKEN);
+        }
+
+        int nextChar = peek();
+        switch (nextChar) {
+            case LexerTerminals.BITWISE_XOR:
+            case LexerTerminals.DOLLAR:
+                startMode(ParserMode.RE_ASSERTION);
+                this.reader.advance();
+                break;
+            case LexerTerminals.DOT:
+                startMode(ParserMode.RE_ATOM);
+                this.reader.advance();
+                break;
+            case LexerTerminals.BACKSLASH:
+                processReEscape(false);
+                break;
+            // Handle "[" ["^"] [ReCharSet] "]".
+            case LexerTerminals.OPEN_BRACKET:
+                this.reader.advance();
+                startMode(ParserMode.RE_ATOM_CHAR_SET);
+                if (peek() == LexerTerminals.BITWISE_XOR) {
+                    this.reader.advance();
+                }
+                return getRegExpText();
```

Comment:
I think we should have a separate node `"[" ["^"] [ReCharSet] "]"` otherwise there is no use of parsing like this right?. We need to do this for other nodes as well like `XLMParser`